p3387.tex

```diff
@@ -128,21 +132,23 @@ \subsection{Coroutine-ness is an implementation detail}
 
-One of the design principles of the Contracts MVP (\cite{P2900R8}, Section 3.1, Design Principle 11) is that \tcode{pre} and \tcode{post} serve both caller and callee. In this case, the callee is the ramp function: \tcode{pre} on a coroutine is an assertion on the parameters passed into the ramp function when it is called, \emph{before} the coroutine state is constructed, and the state of the program at that point; \tcode{post} on a coroutine is an assertion on the object that the ramp function returns to the caller at the point of first suspension, and the state of the program at that point.
+One of the design principles of the Contracts MVP (\cite{P2900R8}, Section 3.1, Design Principle 11) is that \tcode{pre} and \tcode{post} serve both caller and callee. In this case, the callee is the ramp function: \tcode{pre} on a coroutine is an assertion on the parameters passed into the ramp function when it is called, \emph{before} the coroutine state is constructed, and the state of the program at that point; \tcode{post} on a coroutine is an assertion on the object that the ramp function returns to the caller.
 
-Crucially, the callee of \tcode{pre} and \tcode{post} on the coroutine declaration is \emph{not} the coroutine body --- the function body that the user wrote --- and \tcode{pre} and \tcode{post} on the coroutine declaration do not assert anything directly related to that body. The coroutine body may run on another thread, resume long after the postconditions that apply to the declaration are checked, or even finish before then.\footnote{The coroutine state can be destroyed on initial suspend, which happens before \tcode{post} is checked.} Furthermore, the coroutine body never directly sees the object being returned from the function.
+Crucially, the callee of \tcode{pre} and \tcode{post} on the coroutine declaration is \emph{not} the coroutine body --- the function body that the user wrote --- and \tcode{pre} and \tcode{post} on the coroutine declaration do not assert anything directly related to that body. The coroutine body may run on another thread, resume long after the postconditions that apply to the declaration are checked, ??? or even finish before then. \footnote{The coroutine state can be destroyed on initial suspend, which happens before \tcode{post} is checked.} Furthermore, the coroutine body never directly sees the object being returned from the function.
+%IDS I am not sure what conditions you envisage here;  A coroutine could suspend at the initial suspend and then be destroyed before the body was executed - but that would still return the coroutine object to the caller _before_ any such destruction.  Perhaps you envisage some fiendish implementation of an awaiter that could somehow do this? (at the moment I can't think of one - but ...)
```

Comment:
A promise type could ship the coroutine_handle to another thread to finish evaluating and store a pointer to the return object (and maybe a barrier) in what gets shipped.... there doesn't need to be anything in the return object that keeps a reference to the coroutine handle alive, so it can be destroyed before the current thread wakes up and returns control back to the caller when initial_suspend is invoked.

Reply:
My confusion is about *what* is destroyed here

 - (1) the coroutine state 
  -- this cannot be destroyed at the initial awaiter without the coroutine suspending (and therefore returning the coroutine object to the ramp's caller) - this is because it is UB to destroy a running coroutine.

 -- it is, of course, perfectly possible for a coroutine to run synchronously to completion (i.e. all the suspend points flag 'ready').

 - (2) or the ramp return object (which does not even have to exist for a fire & forget void coroutine)
  - the responsibility for destroying this rests with the ramp - and it is only done when the coroutine exits with an exception.

I need more specifics in the "A promise type could ship the coroutine_handle to another thread to finish evaluating and store a pointer to the return object (and maybe a barrier) in what gets shipped"

It is completely agreed that the ramp's return type does not need to contain any reference to the latched coroutine; as said above it does not even need to exist (we can construct a void coroutine).

Reply:
This stuff is way over my head so I suggest that we just remove "or even finish before then" and the associated footnote. Is that OK?

Reply:
I think that's a fine solution - a very complex and fiendish implementation of promise and awaiter is not a useful distraction to the objective here.

Reply:
So maybe we can change
"The coroutine state can be destroyed on initial suspend, which happens before post is checked"
to
"The coroutine state can be destroyed on initial suspend, in which case it may end up happening before post is checked"?

Reply:
> returning from await_suspend, if the coroutine_handle isn't stashed somewhere, will result in the coroutine being destroyed, right? So if your promise type doesn't stash the handle in the return object or pass it anywhere else in await_suspend, as soon as it suspends the coroutine should be destroyed... I think.

I am not following this; await_suspend () is perhaps misnamed - it is a customisation point that allows three different responses to the case that the current coroutine has decided to suspend.

1. void (that is await_suspend() is evaluated only for its side-effects, which can include resuming some other coroutine.
2. It has a true/false return; this allows the customisation point to say "OK you should now resume" - even though we just decided to suspend.
3. It can return a handle to another coroutine to be resumed.

returning from await suspend results in one of three actions then:
1 ... we suspend (and return to the caller) if the caller  is the ramp then that triggers the return of the constructed coroutine object (in all but the pathological void case).
2... if the bool is false - we reenter the dispatch table as if we had been resumed
      if the bools is true we behave as (1)
3 .. we "continue" by resuming the coroutine mentioned in the returned handle (so-called continuation)

> But that is when await_suspend returns, which is immediately before returning to the caller, but it certainly isn't going to be guaranteed to be sequenced after evaluating postcondtions on the ramp function.

Post-conditions on the ramp are (for GCC at least) arranged to be placed on every non-exceptional edge out of the ramp - which includes any edge that is a 'suspend'.

> Here's an example where a promise type never allows resuming after initial suspend : https://godbolt.org/z/Tfn9d9Y53
> 
> A more realistic example would involve race conditions where the handle is shipped off to another thread, resumed, and might be destroyed before await_suspend actually returns to its caller on initial_suspend (and so the coroutine handle is destoryed by then).

There are some concerns in my mind (that are similar the concerns about fibres) and TLS .. 

For this paper - IMO the best solution is not to get into this discussion - but I think it valuable to clarify any edge cases that we implementers might have missed - so happy to continue the discussion here or elsewhere.

Reply:
My understanding is that the point of await_suspend is to do stuff with the coroutine handle when actually suspending, like ship it to another thread where it will be resumed (and potentially destroyed)... so by the time await_suspend returns, the coroutine handle can always potentially be destroyed.  Some promise types, like generator,  might stash the handle in the return object so that that return object can be used to resume the coroutine --- but I don't think that is necessarily the case.

I believe that works and is supported (intentionally), but it means you have to be very careful when reasoning about what happens when the ramp function is executing and when the rest of the coroutine body might execute.

Reply:
> My understanding is that the point of await_suspend is to do stuff with the coroutine handle when actually suspending, like ship it to another thread where it will be resumed (and potentially destroyed)... so by the time await_suspend returns, the coroutine handle can always potentially be destroyed. Some promise types, like generator, might stash the handle in the return object so that that return object can be used to resume the coroutine --- but I don't think that is necessarily the case.

let's try to keep it simple - and describe what happens in single-threaded case without any secondary coroutine...  (modulo concerns about passing coroutines between threads because of TLS point caching) multi-threaded or more complex cases are an extension of this ...

> I believe that works and is supported (intentionally), but it means you have to be very careful when reasoning about what happens when the ramp function is executing and when the rest of the coroutine body might execute.

Agreed we need to be careful - but I *think* we have been - and the coroutine's internal state is separate from the ramp's return object.

 **It is also true that we have a vulnerability in that we cannot assume that a coroutine handle actually has any meaning w.r.t to alive-ness of its contained pointee .. and we have no current mechanism to work around this - which is a PITA .. but this is a general issue with pointers, after all  (I lament this situation, but have no current cunning scheme in mind to improve it)** 

=== anyway ... to try and simplify ... 

what we have is this (where `aw` is an awaiter)

```
if (!aw.await_ready()) {

(??? = ) aw.await_suspend (self_handle);

** what happens here depends on the return type of the await_suspend() (as described above)
    (probably the most common case is the void return, in which case we unconditionally suspend)
*** here is where we return to the caller in the case we suspend (not from within the await_suspend)
*** when the caller is the ramp, then we return the coroutine object (whatever that is) to the ramp's caller.

}

resume:
aw.await_resume();
```

So IIUC your point is that we could (for a simple example) call handle.destroy() on the self_handle from within the await_suspend()?   (which is not UB because we are suspended).

That would destroy the coroutine state - but it does not touch the ramp coroutine return object.

`aw.await_suspend()` returns (with one of the three return cases)

The two (and a half) non-UB cases that would be allowed - 

 1. void await_suspend () - which would continue on the suspend path - returning to the caller; if that is the ramp then we'd still progress the return object and return that to the ramp's caller.   At this point the coroutine is dead - but the coroutine return object is still valid (so if one intends to use it in a post that determines the coroutine is OK -  one better make sure it is setup with a default of false ...)

2. a handle return - in which case we would carry out a continuation on the.

1a) A bool return that is true would also DTRT ..
but a false bool return would be UB because it would try to resume a dead coroutine.

Reply:
or ..
returning a coroutine handle as the ramp return is suitable for compiler test-cases - but a lousy idea for production code - since there's no way to know if it represents a valid coroutine.

Reply:
In any event, I think the note [and the sentence fragment related] is not correct 
- regardless of how the coroutine progresses (either suspends, or possibly runs to synchronous completion and self-destructs - or is destroyed from within await_suspend()) ...

...  the outgoing edge to the ramp will still carry the return object and evaluate the post condition - the onus is on the coroutine object author to ensure that it, and any post checks applied to it, are meaningful (and, of course, the larger frameworks like folly do not return bare handles, but more complex objects that embed handles to do the low-level manipulation).

Reply:
I have to admit I can't follow the technical details of this discussion. Can you tell me how you'd like the paper text to be changed? Thanks

Reply:
I need to stand down for the night - but assuming that Josh and I are on the same page (I *think* we are) I can suggest an edit in the morning.

Reply:
That's fine - but since I'd like to share this paper with SG21 today, if it's OK with you, I'll fold the rest of the PR into the paper now, and share it, and then we can resolve this specific bit later. I hope these comments won't disappear when I merge the PR...

Reply:
OK.. understood - I don't think this stuff will be lost with a merge PR - rebases I'm not so sure.

Reply:
I changed my mind - since there are at least two open issues remaining, I'll leave the PR open until tomorrow, so we can resolve both before we merge the PR and share the paper with SG21. Thanks.

Reply:
(unless @notadragon objects) I suggest that for this paper delete ", or even finish before then" and note1.

In general,  in my suggested edits, I tried to delete all text that made the assumption that the ramp would finish "at first suspension" (because that is not the _**only**_ way it can finish - which I believe is the summary of this discussion) - but I might still have missed some cases.